cwms-http-client/src/main/java/mil/army/usace/hec/cwms/http/client/PreferencesBackedCookieStore.java

```diff
@@ -0,0 +1,145 @@
+                } catch (IOException e) {
+                    LOGGER.log(Level.WARNING, e, () -> "Unable to store preference cookies to cookie manager for URL: " + entry.getKey());
+                }
+            }
+        } catch (BackingStoreException e) {
+            LOGGER.log(Level.CONFIG, e, () -> "Unable to obtain cookies stored in preferences at node: " + preferences.absolutePath());
+        }
+    }
+
+    private synchronized Map<URI, List<String>> extractCookiesFromPreferences() throws BackingStoreException {
+        Map<URI, List<String>> uriToCookie = new HashMap<>();
+        for (String child : preferences.childrenNames()) {
+            Preferences node = preferences.node(child);
+            String uriAddress = node.get(URI_KEY, "");
+            URI uri = URI.create(uriAddress);
+            List<String> cookies = uriToCookie.computeIfAbsent(uri, u -> new ArrayList<>());
+            for (String cookieName : node.keys()) {
+                if (!URI_KEY.equals(cookieName)) {
+                    cookies.add(node.get(cookieName, ""));
+                }
+            }
+        }
+        return uriToCookie;
+    }
+
+    public synchronized void writeCookiesToPreferences() {
+        try {
+            List<URI> urIs = cookieManager.getCookieStore().getURIs();
+            for (URI key : urIs) {
+                try {
+                    for (List<String> value : cookieManager.get(key, new HashMap<>()).values()) {
+                        storeCookiesForUrl(key, value);
+                    }
+                    //The InMemoryCookieStore strips out the 's' in "https" on storage. Need to check cookies for both URI's
+                    if (key.toString().contains("http:")) {
+                        URI https = URI.create(key.toString().replace("http:", "https:"));
+                        for (List<String> value : cookieManager.get(https, new HashMap<>()).values()) {
+                            storeCookiesForUrl(https, value);
+                        }
+                    }
+                } catch (IOException e) {
+                    LOGGER.log(Level.WARNING, e, () -> "Unable to obtain cookies for url: " + key);
+                }
+            }
+            preferences.flush();
+        } catch (BackingStoreException e) {
+            LOGGER.log(Level.WARNING, e, () -> "Unable to write cookies to preferences node: " + preferences.absolutePath());
+        }
+    }
+
+    private void storeCookiesForUrl(URI key, List<String> value) {
+        Preferences node = preferences.node(key.getHost());
+        node.put(URI_KEY, key.toString());
+        for (String cookie : value) {
+            HttpUrl httpUrl = HttpUrl.get(key);
+            if (httpUrl != null) {
+                Cookie parse = Cookie.parse(httpUrl, cookie);
+                if (parse != null) {
+                    node.put((parse.domain() + " | " + parse.name()).toLowerCase(), cookie);
```

Comment:
Since preferences are a key/value pair (that I want to be able to clobber over if I get a new cookie value with the same name and domain), I need a unique key. The entire cookie is serialized as the value so when I restore, I don't need the key. The pipe is just for readability.

My rule of thumb for constants is that I need to use them more than once or at least be able to give them a good name. I was meh on the name for this one. How does DOMAIN_COOKIE_DELIMITER sound to you?